### ISSUE TITLE
Möller–Trumbore intersection algorithm

### ISSUE THREAD

**Issue by Uwila** [recp/cglm#129]
Perhaps the [Möller–Trumbore intersection algorithm](https://en.wikipedia.org/wiki/M%C3%B6ller%E2%80%93Trumbore_intersection_algorithm) would make a good addition to this library? I would be willing to try and implement it; I already made a version of it for a personal project that just uses C and cglm, so it wouldn't be too hard.

I'm not sure if you want a contribution like this though, and I also don't know whether it would be compatible with the license. The Wikipedia page is CC-SA 3.0 (I don't know if in this case a simple notice in a comment would suffice to point out the source, [this is what Wikipedia seems to imply though](https://en.wikipedia.org/wiki/Wikipedia:Reusing_Wikipedia_content#Re-use_of_text_under_Creative_Commons_Attribution-ShareAlike) as long as it would make the result just a modified version and not the entire library by implication a derivative work) but the algorithm originally comes from [this paper](http://www.graphics.cornell.edu/pubs/1997/MT97.pdf) I believe, and I would not know the implications of basing code on a snippet in a paper for the required license.

**Comment by recp** [comment]
@Uwila perfect! It would be nice if there is no license penalty. 

It seems **glm** also implemented Ray Triangle Intersect [glm/gtx/intersect.inl](https://github.com/g-truc/glm/blob/ea678faff9340ae4a79f50f2edd947141405e128/glm/gtx/intersect.inl) It may used same algorithm or code snippets

**Comment by Uwila** [comment]
@recp I've started implementing it, but am running into an issue. I think the inline version works now, but when trying to use the pre-compiled version I get `main.c:(.text+0x25f0): undefined reference to `glmc_line_triangle_intersect'` in my test project.

I'm new to building libraries in C so I'm thinking you might see something obvious that I'm missing. You can look at what I have so far [here](https://github.com/Uwila/cglm/commit/486e3826b1ac24297a73ce62c6c61f3ff207e1a8). If not, I'll probably figure it out eventually.

Aside from that I have a question about taste: The algorithm computes whether there's a collision, and what the distance to the point of collision is. Right now it returns `-1.0f` when there's no collision. I could make it so that instead of returning the distance, it just returns a boolean, and to get the distance you have to pass a pointer to a float for the function to save the distance in. What would have your preference?

**Comment by recp** [comment]
Hello @Uwila 

Thanks for starting work on this 🤗

---

> main.c:(.text+0x25f0): undefined reference to `glmc_line_triangle_intersect' in my test project.

Did you link **cglm** to your test project? Which platform do you test on? For instance in **clang**, **gcc** family you add `-lcglm` flag to specify that you are linking/adding a library to your project.`-l`(L) means library, and `-I` (i) means include.

For windows must must see Linker options in Visual Studio. Also it seems you did not add `ray.c` to Visual Studio project, so this file is not compiled on windows...

---

in **glm** there are a few more functions, I don't know what one does what for now:

https://glm.g-truc.net/0.9.4/api/a00179.html
[Github -> glm/gtx/intersect.hpp](https://github.com/g-truc/glm/blob/master/glm/gtx/intersect.hpp)
[Github -> glm/gtx/intersect.inl](https://github.com/g-truc/glm/blob/ea678faff9340ae4a79f50f2edd947141405e128/glm/gtx/intersect.inl)

```C++
bool intersectLineSphere(); /* is this related to ray / ray tracing? */
bool intersectLineTriangle(); /* is this related to ray / ray tracing? */
bool intersectRayPlane();
bool intersectRaySphere();
bool intersectRayTriangle();
```

`glm_line_triangle_intersect()` name seems very long to me. 

For instance in **cglm** all of these are intersection functions:

```C
bool glm_aabb_aabb(vec3 a[2], vec3 b[2]);  /* check if two AABB intersects */
bool glm_aabb_sphere(vec3 box[2], vec4 s); /* check if AABB intersects with sphere */
bool glm_aabb_point(vec3 box[2], vec3 p);  /* check if point is inside of AABB */
bool glm_sphere_sphere(vec4 s1, vec4 s2);  /* check if two sphere intersects */
```

I'm not sure users like it or not.

May I ask what is the difference between `intersectLineTriangle` and `intersectRayTriangle`, if do you know?

Maybe we kan group **ray** related function inside `glm_ray_`, what do you think?

```C
bool glm_ray_intersect_plane();
bool glm_ray_intersect_sphere();
bool glm_ray_intersect_triangle();
```

```C
/* intersect is intended */
bool glm_ray_plane();
bool glm_ray_sphere();
bool glm_ray_triangle();
```

---

> I could make it so that instead of returning the distance, it just returns a boolean, and to get the distance you have to pass a pointer to a float for the function to save the distance in. What would have your preference?

I like this idea. 👍

I want to implement Ray Tracing in my render engine in the future (https://github.com/recp/gk) these features may help me a lot :)

**Comment by Uwila** [comment]
> Did you link cglm to your test project?

I did, I tested it by using another `glmc` function which works fine. I'm thinking there's probably something I'm forgetting that's required to compile it into the library. For example, in `Makefile.am` I've added `include/cglm/ray.h` to `cglm_HEADERS`, `include/cglm/call/ray.h` to `cglm_call_HEADERS`, and `src/ray.c` to `libcglm_la_SOURCES`. Also, I've added `#include "call/ray.h"` in `include/cglm/call.h`. I'm thinking there might be another place I've forgotten to add the new files.

--- 

On the naming: I think basically all functions that involve a ray and a shape will be to check for intersection. So I think your last proposal (`glm_ray_triangle` and the like) will work fine.

> May I ask what is the difference between intersectLineTriangle and intersectRayTriangle, if do you know?

I don't know for sure, but my guess would be that a ray has some start point and goes on forever in one direction from there, while a line goes on forever in both directions.

**Comment by recp** [comment]
> I did, I tested it by using another glmc function which works fine.
> (See the commit in my previous comment for all changes) I'm thinking there might be another place I've forgotten to add the new files.

I saw your commit it seems fine except mis placed indents. Maybe you should call `make clean && make`. Did you get error on compile time or runtime? If runtime then make sure that you have a copy of cglm library in correct folder. If you use MSVC then ray.c must be added to VS project file too. 

I have downloaded your branch and it compiled successfully, also I called your new function from test target then it did not complain. 

```C
#include "src/test_common.h"

int
main(int argc, const char * argv[]) {
  ...

  vec3 origin, direction, v0, v1, v2;
  
  test_rand_vec3(origin);
  test_rand_vec3(direction);
  test_rand_vec3(v0);
  test_rand_vec3(v1);
  test_rand_vec3(v2);

  printf("before glmc_line_triangle_intersect()\n");
  glmc_line_triangle_intersect(origin, direction, v0, v1, v2);
  printf("after glmc_line_triangle_intersect()\n");

  ...
}
```

this works (make check).

---

> So I think your last proposal (glm_ray_triangle and the like) will work fine.

nice 🤗

> I don't know for sure, but my guess would be that a ray has some start point and goes on forever in one direction from there, while a line goes on forever in both directions.

thanks

**Comment by Uwila** [comment]
I found out what the problem was: I had previously installed cglm using my package manager, and while working on it I used `make install`, which installed into a different directory; so I was indeed linking succesfully, just to the old version.

I think I have everything working now, including documentation. I don't have a Windows setup for programming at all so the changes to the `win/` directory are not tested by me.

On the license issues: I'm not really sure what the case is with this stuff. I see this same algorithm used all over the place (the glm stuff looks similar to the paper implementation in my opinion, and I see the paper implementation elsewhere on GitHub too, all with different licenses. There is a link to the source in the original paper, which may come with a license, but this is a dead link), so I suppose using it is fine, there's a lot of precendent for that :smile: I don't know whether "use it until someone complains" is a bad idea but I wouldn't know what else to do.

So I hope that's all fine and I'll have my pull-request coming in shortly. If anything needs changing just let me know, of course.
